src/lib/alloc.c

```diff
@@ -703,6 +722,48 @@ void rfree(void *ptr)
 	memmap.heap_trace_updated = 1;
 }
 
+void *_realloc(void *ptr, int zone, uint32_t caps, size_t bytes, bool copy)
+{
+	void *new_ptr = NULL;
+	uint32_t flags;
+
+	/* sanity check */
+	if (!ptr)
+		return ptr;
+
+	spin_lock_irq(&memmap.lock, flags);
+
+	new_ptr = _malloc_unlocked(zone, caps, bytes);
+
+	if (new_ptr && copy)
+		memcpy_s(new_ptr, bytes, ptr, bytes);
+
+	spin_unlock_irq(&memmap.lock, flags);
```

Comment:
Shouldn't we free the ptr?

Reply:
@tlauda, well @lgirdwood just pointed in previous comments _not_ to free it. It seems to be a little bit ambiguous from the definition/history. Old c-hackers in the office here seems to use it so that realloc doesn't free the ptr and you have to call free yourself. As I interpret it, the functionality is implementation specific, and whatever we decide I'm fine.

Reply:
C standard says that `ptr` should be *invalidated*. You can do with it whatever is needed in your allocator, but in the end user shouldn't use ptr after realloc.

Reply:
@jajanusz @tlauda updated

Reply:
@juimonen There is one more thing that is not aligned with standard. `realloc` is like `malloc` when you pass null, `_rfree_unlocked` can handle null ptr, but memcpy_s will fail.

Reply:
ok, sorry missed that.

Reply:
@juimonen you need to free it if it's been moved, from the man page
```If the area pointed to was moved, a free(ptr) is done.```

Reply:
@lgirdwood ok. In this case I'm always allocating new area. The in place shrink or enlarging could be done (I guess), but it looked quite complicated and error prone for me to touch...

Reply:
@juimonen yep, I agree new area is always best which means we should always free ptr locally (and not done by caller)